Fix `Object::notification` order

Sauermann opened the pull request:
- resolve #52325
- resolve godotengine/godot-proposals#5585
- CPP-bindings: godotengine/godot-cpp#1151

Previously the `p_reversed` parameter didn't influence the order in a correct way.
Also script overridden _notification functions were not called in the correct order.

To fix this, some `notification` functions had to add a `p_reversed` parameter. This made it necessary to adjust mono and cpp bindings, since they use these functions.

MRP for testing: [ReverseNotification.zip](https://github.com/godotengine/godot/files/11854113/ReverseNotification.zip) (based on notes in #52325)

- [X] Investigate necessary changes for GDExtension (MRP: [ReverseNotificationCPP.zip](https://github.com/godotengine/godot/files/12310702/ReverseNotificationCPP.zip))
- [X] Investigate necessary changes for mono (MRP: [ReverseNotificationMono.zip](https://github.com/godotengine/godot/files/11856493/ReverseNotificationMono.zip))
- [X] Ensure compatibility (see [comment](https://github.com/godotengine/godot/issues/63275#issuecomment-1666360452))
- [X] Revert debugging print_line additions

Updated 2023-08-29: Fix merge conflict with #80394

Sauermann committed:
Fix Object::notification order

Previously the `p_reversed` parameter didn't influence the order
in a correct way.
Also script overridden _notification functions were not called in
the correct order.

To fix this some `notification` functions had to add a `p_reversed`
parameter.

This made it necessary to adjust cpp-bindings.

Co-authored-by: David Snopek <dsnopek@gmail.com>

Maran23 commented:
IMO we should consider writing tests for this as well (either in this PR or in a follow up).

aaronfranke commented:
This PR is failing the CI checks, those need to be resolved.

dsnopek commented on core/object/object.cpp:
```diff
@@ -795,14 +795,24 @@ Variant Object::call_const(const StringName &p_method, const Variant **p_args, i
 
-	if (script_instance) {
-		script_instance->notification(p_notification);
+	if (_extension && _extension->notification2) {
+		_extension->notification2(_extension_instance, p_notification, p_reversed);
```
In order for old GDExtensions to continuing working, we'll need to still call `_extension->notification` if `_extension->notification2` is not set. So, something like:

```suggestion
	if (_extension) {
		if (_extension->notification2) {
			_extension->notification2(_extension_instance, p_notification, p_reversed);
		} else if (_extension->notification) {
			_extension->notification(_extension_instance, p_notification);
		}
```

dsnopek commented on core/extension/gdextension_interface.h:
```diff
@@ -287,6 +287,29 @@ typedef struct {
+	GDExtensionClassFreePropertyList free_property_list_func;
+	GDExtensionClassPropertyCanRevert property_can_revert_func;
+	GDExtensionClassPropertyGetRevert property_get_revert_func;
+	GDExtensionClassNotification notification_func;
+	GDExtensionClassNotification2 notification_func2;
```
I think we should add a comment here explaining that `notification_func` should be NULL, and is only provided for backward compatibility.

dsnopek commented on core/extension/gdextension.cpp:
```diff
@@ -337,6 +337,65 @@ void GDExtension::_register_extension_class(GDExtensionClassLibraryPtr p_library
 	ClassDB::register_extension_class(&extension->gdextension);
 }
+
+void GDExtension::_register_extension_class2(GDExtensionClassLibraryPtr p_library, GDExtensionConstStringNamePtr p_class_name, GDExtensionConstStringNamePtr p_parent_class_name, const GDExtensionClassCreationInfo2 *p_extension_funcs) {
```
Rather than maintaining both `_register_extension_class()` and `_register_extension_class2()` with nearly identical code, I think `_register_extension_class()` should be made to build the new `GDExtensionClassCreationInfo2` struct and call `_register_extension_class2()`.

dsnopek commented on core/object/script_language_extension.h:
```diff
@@ -737,10 +737,15 @@ class ScriptInstanceExtension : public ScriptInstance {
+	}
+
+	virtual void notification(int p_notification, bool p_reversed = false) override {
+		if (native_info->notification_func2) {
+			native_info->notification_func2(instance, p_notification, p_reversed);
```
Again, this'll need to still call `notification_func` if `notification_func2` isn't set, so that older GDExtensions will keep working:

```suggestion
		if (native_info->notification_func2) {
			native_info->notification_func2(instance, p_notification, p_reversed);
		} else if (native_info->notification_func) {
			native_info->notification_func(instance, p_notification);
```

dsnopek reviewed (changes requested):
The GDExtension compatibility changes are definitely headed in the right direction! I have just a couple of notes

Sauermann commented:
@dsnopek Thank you for the detailed review. I believe, that I have addressed all your comments.

From other deprecation-uses, I'm aware of
```c++
#ifdef DISABLE_DEPRECATED
#endif // DISABLE_DEPRECATED
```

Should they be used here as well?

dsnopek commented:
> Thank you for the detailed review. I believe, that I have addressed all your comments.

Thanks, the changes look great!

> From other deprecation-uses, I'm aware of
> 
> ```c++
> #ifdef DISABLE_DEPRECATED
> #endif // DISABLE_DEPRECATED
> ```
> 
> Should they be used here as well?

Hm. I'm not exactly sure how we'd use that in this context.

`gdextension_interface.h` isn't a "normal" part of Godot's source code, it's meant to be used by other projects for interfacing with Godot, and may not even be parsed by a C/C++ compiler (as is the case for the godot-rust bindings for example), making it almost more like data than code. So, in the past we've made a point of using only the absolute minimum of preprocessor directives in `gdextension_interface.h`.

But maybe we could exclude the implementation of the old function in `gdextension.cpp` if `DISABLE_DEPRECATED` is defined? I think that's likely to break binary compatibility for plenty of GDExtensions, though.

Sauermann commented:
Thanks for the explanation, which lets me think, that `DISABLE_DEPRECATED` shouldn't be used here.

dsnopek reviewed (approved):
I just tested this locally together with the godot-cpp PR, and it worked great! I also tried loading an older GDExtension using Godot built with this PR, and it also worked (so, using the compatibility code). And the code here matches what we've discussed at GDExtension meetings with regard to compatibility for this sort of situation.

So, in short, from the perspective of GDExtension, this all looks great!

Bromeon commented on core/extension/gdextension_interface.h:
```diff
@@ -287,6 +287,29 @@ typedef struct {
 	void *class_userdata; // Per-class user data, later accessible in instance bindings.
 } GDExtensionClassCreationInfo;
 
+typedef void (*GDExtensionClassNotification2)(GDExtensionClassInstancePtr p_instance, int32_t p_what, bool p_reversed);
```
Would it make sense to list this immediately below `typedef void (*GDExtensionClassNotification)`, so the differences are obvious?

Sauermann replied:
I can change that if needed. Would like to hear an additional opinion, which variant is easier to read. My original thought was to group all new typedefs together.

Gallilus replied:
I think https://github.com/godotengine/godot/pull/76577/files is a very clean way of doing it.
Only just realised it is a draft so. @reduz this should be possible, not? @RedworkDE mentions it requires [76446](https://github.com/godotengine/godot/pull/76446) but that is meged.

dsnopek replied:
> I can change that if needed. Would like to hear an additional opinion, which variant is easier to read. My original thought was to group all new typedefs together.

Personally, I don't have a strong preference, I think either way is fine. But I'd maybe _slightly_ lean to towards having this `typedef` by the other function pointer `typedefs` :-)

Sauermann replied:
I have moved the `typedef` there.

Bromeon commented on core/extension/gdextension_interface.h:
```diff
@@ -287,6 +287,29 @@ typedef struct {
+	GDExtensionClassFreePropertyList free_property_list_func;
+	GDExtensionClassPropertyCanRevert property_can_revert_func;
+	GDExtensionClassPropertyGetRevert property_get_revert_func;
+	GDExtensionClassNotification notification_func; // Deprecated - should be nullptr. Only kept for backward compatibility.
```
The comment is not really clear -- "should be nullptr", why is it there then? Under which circumstances does an extension developer need to provide this value?

If someone uses `classdb_register_extension_class2`, that sets their minimum Godot version to 4.2; so they cannot support older extensions -- so the compat method wouldn't need to be necessary?

Sauermann replied:
I have updated the comment and made it more clear.

If someone uses in Godot 4.2 `classdb_register_extension_class`, they still need `notification_func` to be available, so it is necessary, to keep this available.
Please see https://github.com/godotengine/godot/pull/78634#discussion_r1287495323 for why I added the comment.

Bromeon replied:
Thanks 🙂 

> If someone uses in Godot 4.2 `classdb_register_extension_class`, they still need `notification_func` to be available, so it is necessary, to keep this available.

But `classdb_register_extension_class` accepts a different struct: `GDExtensionClassCreationInfo`.

Why `GDExtensionClassCreationInfo2` needs the field `notification_func` field is still not clear to me. That struct can only be used with `classdb_register_extension_class2`, which sets the API minimum version to 4.2. At this point one should _always_ use the `notification_func2`, if I understand correctly.

Maybe it's because of @dsnopek's [comment here](https://github.com/godotengine/godot/pull/78634#discussion_r1287496694) suggesting to reduce code duplication? In that case, I'd argue this is an 
implementation detail that should **not** leak to the user-facing API, because it has no valid use case.

Maybe I'm missing something, but I'd probably remove that field and have a
```cpp
GDExtensionClassNotification2 notification_func;
```
instead, even if it means the C++ code has a bit of duplication.

@Sauermann @dsnopek what do you think?

dsnopek replied:
Yeah, we could probably get rid of `notification_func` (or keep the name, but change the type) on the new  `GDExtensionClassCreationInfo2` struct, because it's not really needed in the interface.

That would make the implementation of `_register_extension_class()` a little more complicated, though, as currently it builds the new struct and calls `_register_extension_class2()`. We'd need to either go back to duplicating code in `_register_extention_func()` and `_register_extension_func2()`, or come up with a way to refactor them so they both used a shared 3rd function, or something like that?

Internally, we still need to keep both `notification_func` and `notification_func2` on the `ObjectGDExtension` struct, so I don't think it's the worst thing for them to both be on the interface struct too? I guess it's a question of how much we want to try and hide the internals here.

dsnopek replied:
Ack, @Bromeon you and I cross posted :-) Yeah, I think we're both saying the same thing

Bromeon replied:
Thanks for confirming! 

I think the Godot C++ implementation (what is stored inside `native_info`) and the GDExtension API will likely diverge _more_ over time, as new features are provided.

With the current approach, that would mean the new API needs to tag along all possible features from previous versions, even though they no longer make sense for an extension developer. Now it's just 1 field, but if we ever have `GDExtensionClassCreationInfo3`, we'd _still_ need to keep around `notification_func` from v1, plus whatever was deprecated in v2.

---

Regarding implementation, this option would be quite nice duplication-wise:
```cpp
struct ClassCreationInfo {
   // All the new stuff
   GDExtensionClassNotification2 v2;
   
   // Legacy features as single fields
   GDExtensionClassNotification notification_func1;
};
```

dsnopek replied:
This conversation made me realize we missed a compatibility issue around the `GDExtensionScriptInstanceInfo` struct! I made a new comment about it below.

> I think the Godot C++ implementation (what is stored inside native_info) and the GDExtension API will likely diverge more over time, as new features are provided.
>
> With the current approach, that would mean the new API needs to tag along all possible features from previous versions, even though they no longer make sense for an extension developer. Now it's just 1 field, but if we ever have `GDExtensionClassCreationInfo3`, we'd _still_ need to keep around `notification_func` from v1, plus whatever was deprecated in v2.

Personally, I don't think having a few deprecated fields accruing over time is the worst thing. Maintaining compatibility is always going to lead to some ugliness/messiness somewhere, and it's really just a matter of where you want the ugliness to live.

The current PR sort of shares the ugliness equally with the interface (`gdextension_interface.h`) and the internal Godot implementation. If we prioritize making `gdextension_interface.h` as clean as possible, that means increasing the ugliness in the internal Godot implementation. In particular, I think the steps to do this look something like:

1. Make a new struct based on `GDExtensionScriptInstanceInfo` but only for internal use by `ScriptInstanceExtension`, which has both notification fields. Godot would copy the data from `GDExtensionScriptInstanceInfo` or `GDExtensionScriptInstanceInfo2` for use internally.
2. We don't need to do the same thing for `GDExtensionClassCreationInfo` because we already have `ObjectGDExtension` for use internally
3. `GDExtensionScriptInstanceInfo` and `GDExtensionScriptInstanceInfo2` as well as `GDExtensionClassCreationInfo` and `GDExtensionClassCreationInfo` would each only have one `notification_func` - the old one on the old structs, and the new one on the new structs.
4. Refactor or duplicate the code in `_register_extension_class()`/`_register_extension_class2()` and `gdextension_script_instance_create()`/`gdextension_script_instance_create2()`, to accept the interface structs and copy the data on to the internal structs, rather than just having the older function construct the newer struct and pass it to that function.

This is all totally doable! But it is a bunch of extra code to maintain.

The question is: do we prioritize keeping `gdextension_interface.h` as clean as possible?

Bromeon replied:
Thanks for elaborating the precise effort needed in the implementation! You clearly know that part better than me 🙂 



> If we prioritize making `gdextension_interface.h` as clean as possible, that means increasing the ugliness in the internal Godot implementation. [...]
> The question is: do we prioritize keeping `gdextension_interface.h` as clean as possible?

My personal opinion is that I would trade off implementation complexity, if it means that the API can be nicer. A few reasons for that:

* Typically, the number of people coming in contact with the API is much bigger than those with the implementation, so improvements have a greater impact. Also, API users are less familiar with the matter than maintainers.
* APIs are set in stone once published, while implementation can always be refactored. In particular, other requirements (maybe Godot-internal ones, maybe motivated by new GDExtension features) can prompt a code reorganization.
* The GDExtension API in particular has quite a big surface and is rather complex due to its low-level nature. The fewer pitfalls or unusual constraints there are (like "please always provide `NULL` for this pointer"), the easier and less frustrating it becomes to learn it. 

Of course, I have a biased view as I'm mostly a consumer of GDExtension, but maybe _because_ I went through a lot of the early hiccups, I appreciated well-defined semantics 😉 and I think the header has gone great lengths to this already, especially in 4.1! But I definitely don't want to diminish efforts required to maintain the Godot internals; you can probably judge that aspect better than me 🙂 

---



> 1. Make a new struct based on `GDExtensionScriptInstanceInfo` but only for internal use by `ScriptInstanceExtension`, which has both notification fields.

Here, would my previous suggestion work? Then you can always use the latest API struct and only keep the legacy fields separate:
```cpp
struct ClassCreationInfo {
   // All the new stuff
   GDExtensionClassCreationInfo2 latest;
   
   // Legacy features as single fields
   GDExtensionClassNotification legacy_notification_func;
};
```



> 4\. Refactor or duplicate the code in `_register_extension_class()`/`_register_extension_class2()` and `gdextension_script_instance_create()`/`gdextension_script_instance_create2()`, to accept the interface structs and copy the data on to the internal structs (rather than just having the older function construct the newer struct and pass it to that function)

This should also be simpler with the above, and not even need more duplication, just 1 indirection:
```cpp
// Add new internal method, with 1 extra parameter:
void GDExtension::_register_extension_class_internal(
        ...,
        const GDExtensionClassCreationInfo2 *p_extension_funcs,
        GDExtensionClassNotification legacy_notification_func,
) {
	// Do the actual registration
	GDExtension *self = reinterpret_cast<GDExtension *>(p_library);
	...
	extension->gdextension.library = self;
	extension->gdextension.parent_class_name = parent_class_name;
	extension->gdextension.class_name = class_name;
	...
}

// Registration v1
void GDExtension::_register_extension_class(...,  const GDExtensionClassCreationInfo *p_extension_funcs) {
	// First, copy everything, as it is done now. Leave notification_func NULL.
	const GDExtensionClassCreationInfo2 class_info2 = { ... };

	// Delegate to internal, providing a legacy pointer.
	_register_extension_class_inner(..., &class_info2, p_extension_funcs->notification_func);
}

// Registration v2
void GDExtension::_register_extension_class2(...,  const GDExtensionClassCreationInfo2 *p_extension_funcs) {
	// Delegate to internal, providing NULL for the legacy pointer.
	_register_extension_class_inner(..., p_extension_funcs, NULL);
}
```

dsnopek replied:
> Here, would my previous suggestion work? Then you can always use the latest API struct and only keep the legacy fields separate

Yes, I think that would work. However, at this point, I think the most important thing would be to get alignment on priorities. I'd really love to hear what other GDExtension team members think about this too! I put a "call for comments" in RocketChat, but this may end up needing to be a topic at a GDExtension meeting.

Bromeon replied:
Makes total sense! 👍

dsnopek replied:
As mentioned below, at the GDExtension meeting we decided to prioritize keeping `gdextension_interface.h` clean, so I'm going to mark this conversation as resolved

dsnopek commented on core/extension/gdextension_interface.h:
```diff
@@ -403,6 +427,7 @@ typedef struct {
 
 	GDExtensionScriptInstanceCall call_func;
 	GDExtensionScriptInstanceNotification notification_func;
+	GDExtensionScriptInstanceNotification2 notification_func2;
```
I failed to notice this in my previous review, but adding a field to `GDExtensionScriptInstanceInfo` will also cause compatibility issues. We'll need to do something similar to what has already been done in this PR for `GDExtensionClassCreationInfo`, including adding a new struct and a new function (maybe `script_instance_create2`?) to accept it.

Sorry for not catching this earlier!

Sauermann replied:
In the latest version I believe that I have addressed this: https://github.com/godotengine/godot/compare/36d8b2d8dc68fd8a3d1d83be589a12642d53506f..c4537d7d693f17fed41329d8bd86e4c4f36cd61e
Had to rebase the PR because of a merge conflict.

dsnopek replied:
Thanks! I haven't tested it, but skimming the code, this looks good to me :-)

I guess we just need to resolve on the GDExtension team the question of leaving the deprecated property on the struct in the interface.

reduz commented on modules/mono/csharp_script.h:
```diff
@@ -281,6 +281,8 @@ class CSharpInstance : public ScriptInstance {
 
 	void notification(int p_notification) override;
 	void _call_notification(int p_notification);
+	void notification(int p_notification, bool p_reversed = false) override;
```
Why not remove the older function and make the parameter optional?

reduz commented on modules/gdscript/gdscript.h:
```diff
@@ -330,6 +330,7 @@ class GDScriptInstance : public ScriptInstance {
 	Variant debug_get_member_by_index(int p_idx) const { return members[p_idx]; }
 
 	virtual void notification(int p_notification);
+	virtual void notification(int p_notification, bool p_reversed = false);
```
Code does not make sense, why not remove the old function?

Gallilus replied:
Is it not required to be able to communicate with old extensions?

Sauermann replied:
I believe, that I have addressed this (and also https://github.com/godotengine/godot/pull/78634#discussion_r1298122129) in update https://github.com/godotengine/godot/compare/dfebd2b671e69d5142595265eded82949dafaa1e..36d8b2d8dc68fd8a3d1d83be589a12642d53506f

I'm not sure, how I could apply `ClassDB::bind_compatibility_method` of #80168 to this case, because, this `notification`-function is not registered in `ClassDB`. So this change doesn't cause any compaitbiility-breakage-message in the github-checks.

reduz commented:
Looks pretty good! Save for the comments I added and the GDExtension pending changes.

Gallilus commented:
I was confused as the deprecated function names are unnumbered and the names that need to stay are numbered.
So created an overview.
![PR78634 as is](https://github.com/godotengine/godot/assets/20573784/2c082ffa-498f-4e04-a6c3-042c39a6bd41)
What I would suggest is to name and structure this PR is similar to https://github.com/godotengine/godot/pull/76577/files
Whereas much as possible the deprecated code is stuffed in `.compat.inc`.
The final code implements a standard naming convention without numbers.
And all checks and references that still need to be in the other files are wrapped in `#ifndef DISABLE_DEPRECATED`
Resulting in the following overview.
![PR78634 suggestion](https://github.com/godotengine/godot/assets/20573784/94253b73-3343-44e7-965c-23524d6e1b42)

Gallilus commented:
Where I am struggling with is the double implementation of `GDExtensionClassNotification` in `Object::ObjectGDExtension` and `GDExtensionScriptInstanceNotification` in  `GDExtensionScriptInstanceInfo`
I do not have any idea what the expectations are if using ifdef in a struct.
Maybe renaming these to ‘…_depreciated’ is a solution?

dsnopek commented:
> What I would suggest is to name and structure this PR is similar to https://github.com/godotengine/godot/pull/76577/files
> Whereas much as possible the deprecated code is stuffed in `.compat.inc`.

While this is how it's done for compatibility methods registered in `ClassDB`, I don't know that this can be applied to changes in the GDExtension interface (ie. `gdextension_interface.h`).

Like I mentioned above, we don't want to put any more preprocessor macros into `gdextension_interface.h`, beyond the bare minimum, since it may not be processed by a C/C++ compiler, but instead some tool to generate C binding code for another language.

We could conceivably remove the implementations of the deprecated functions when `DISABLE_DEPRECATED` is defined, but this would break binary compatibility for GDExtensions (possibly in a way that leads to crashes or unexpected behavior) which I'm not sure is something we want to do even when `DISABLE_DEPRECATED` is defined? Maybe someone who knows more about the goal of `DISABLE_DEPRECATED` could speak to that?

akien-mga commented:
> We could conceivably remove the implementations of the deprecated functions when `DISABLE_DEPRECATED` is defined, but this would break binary compatibility for GDExtensions (possibly in a way that leads to crashes or unexpected behavior) which I'm not sure is something we want to do even when `DISABLE_DEPRECATED` is defined? Maybe someone who knows more about the goal of `DISABLE_DEPRECATED` could speak to that?

`DISABLE_DEPRECATED` is meant for people who'd want to remove all compatibility code to only get and use the latest APIs, thus (marginally) reducing the size of the engine. And it doubles as a nice way to find compatibility code in the codebase for eventual cleanup when we change major version and can remove compat code.

So breaking binary compatibility would be the expectation here (all ClassDB compat methods have been registered with `#ifndef DISABLE_DEPRECATED`).

Now, that doesn't necessarily mean it's good for `gdextension_interface.h` specifically, which has other constraints and can use different rules.

Sauermann commented:
> the same thing should be done for `script_instance_create`

That is done in https://github.com/godotengine/godot/compare/c3dfa8fa76c309adbd207635fb9eedd075dcce49..60d88571d73ce334580f5ed1176e37f7a063b1cf

My understanding is, that the function `script_instance_create` can't be removed, because it is declared in `gdextension_interface.h` and that file should not have any `#ifndef`.
That is why I replaced its function-content by a minimal implementation when `DISABLE_DEPRECATED` is defined.

dsnopek commented:
> My understanding is, that the function `script_instance_create` can't be removed, because it is declared in `gdextension_interface.h` and that file should not have any `#ifndef`.

Right, we can't put any `#ifndef` in `gdextension_interface.h`, but it should be OK to completely `#ifndef` away the implementation from `gdextension_interface.cpp` (the .cpp file), because the header only declares function pointers, not the actual functions.

> Also I have verified that the line can't be removed, because it causes a compiler-error (function `script_instance_create` is unused):

I think if you wrap the whole `gdextension_script_instance_create()` function (not just the function body) in `#ifndef DISABLE_DEPRECATED`, then removing this should work.

And having the function fail to load (because it's not registered) is a better way to break compatibility than having the function turn into a no-op, which could lead to crashes or weird behavior (as opposed to the GDExtension just failing to load).

Sauermann commented:
> I think if you wrap the whole `gdextension_script_instance_create()` function (not just the function body) in `#ifndef DISABLE_DEPRECATED`, then removing this should work.

Yes, that did the trick. Thank you for the explanation.

dsnopek commented:
Discussed at the GDExtension meeting, and the consensus among those present was that we should prioritize the cleanliness of the interface, and remove the deprecated fields from the new structs. We also talked about the naming scheme of adding an integer to the new function pointer typedefs and structs, and for the most part we think this is fine, or at least we can't think of a better way that maintains source compatibility.

dsnopek commented:
> Yes, that did the trick. Thank you for the explanation.

Thanks, the change looks good!

> Discussed at the GDExtension meeting, and the consensus among those present was that we should prioritize the cleanliness of the interface, and remove the deprecated fields from the new structs. 

To follow-up on the decision from last night's GDExtension meeting, these are the concrete steps (copied from an earlier post) to implement this:

> * Make a new struct based on `GDExtensionScriptInstanceInfo` but only for internal use by `ScriptInstanceExtension`, which has both notification fields. Godot would copy the data from `GDExtensionScriptInstanceInfo` or `GDExtensionScriptInstanceInfo2` for use internally.
> * We don't need to do the same thing for `GDExtensionClassCreationInfo` because we already have `ObjectGDExtension` for use internally
> * `GDExtensionScriptInstanceInfo` and `GDExtensionScriptInstanceInfo2` as well as `GDExtensionClassCreationInfo` and `GDExtensionClassCreationInfo2` would each only have one `notification_func` - the old one on the old structs, and the new one on the new structs.
> * Refactor or duplicate the code in `_register_extension_class()`/`_register_extension_class2()` and `gdextension_script_instance_create()`/`gdextension_script_instance_create2()`, to accept the interface structs and copy the data on to the internal structs (rather than just having the older function construct the newer struct and pass it to that function)

@Sauermann Sorry for all the back and forth on this! This is the first time we've modified a struct in `gdextension_interface.h` and so there's some amount of figuring things out as we go. When future contributors need to do the same thing, they'll have the model setup in this PR to follow, but as the PR that's setting the model, this has unfortunately meant work (and re-work) for you. Thanks for all your efforts on this! :-)

Sauermann commented:
> Sorry for all the back and forth on this

No need to worry about that. If I had wanted to abandon this PR, I would have already done so. But yeah, originally I thought, this would be a small fix :-D Thanks for taking the time to provide guidance. It is better to work out the issues as early as possible.

> Make a new struct based on `GDExtensionScriptInstanceInfo` but only for internal use by `ScriptInstanceExtension`, which has both notification fields. Godot would copy the data from `GDExtensionScriptInstanceInfo` or `GDExtensionScriptInstanceInfo2` for use internally.

I did deviate from this suggestion in my implementation in order to reduce code duplication.

The rest of the implementation is now also done.

dsnopek reviewed (commented):
Thanks! This is looking really great. I only have some minor notes about naming.

I also tested together with the godot-cpp PR, and everything seems to be working fine!

> I did deviate from this suggestion in my implementation in order to reduce code duplication.

Looks awesome! That's a great way to avoid duplicating code.

Sauermann commented:
Thanks for the comments, I have made the necessary renames.

dsnopek reviewed (approved):
Thanks! From the perspective of GDExtension, this looks ready to me :-)

~~EDIT: There's some debug code still in the PR that'll need to be removed before this can be merged. I'm assuming it'll get removed soon given that it's on the checklist in the description, but putting this note here for production team folks who might be feeling merge-y :-)~~

EDIT 2: The debug code has now been removed!

Bromeon reviewed (approved):
Thanks a lot! And cool to see the `_internal` approach worked out in reducing code duplication! 🙂

Sauermann commented:
Debug code is now removed.

@Bromeon Thanks for your idea of reducing code-redundancy in https://github.com/godotengine/godot/pull/78634#discussion_r1297714654!

akien-mga commented on core/extension/gdextension_interface.h:
```diff
@@ -2119,13 +2180,28 @@ typedef void (*GDExtensionInterfaceRefSetObject)(GDExtensionRefPtr p_ref, GDExte
  *
  * Creates a script instance that contains the given info and instance data.
  *
+ * This is deprecated. Use `script_instance_create2` instead.
```
Should we have an attribute for that, that also conveys the information of when this was first deprecated?

Bromeon replied:
Great idea, would also make it machine-parseable.

Something that includes both version number when it was deprecated, and a message?
Just an idea:
```cpp 
/*
 * @deprecated 4.2 "Use `script_instance_create2` instead."
 */
```

Syntax can of course be bikeshedded, quotes may not be necessary. Would be allowed to span multiple lines, I guess.

dsnopek replied:
The doc comments in `gdextension_interface.cpp` are based on Doxygen which defines `@deprecated` (an alias for `\deprecated`) this way: https://www.doxygen.nl/manual/commands.html#cmddeprecated

If we follow that pattern, all the text after `@deprecated` is part of a paragraph that ends when the first blank comment line is encountered.

Perhaps we could use a sort of narrative style (where it's assumed the text starts with "Deprecated ...") like:

```c++
/**
 * @deprecated in Godot 4.2. Use `script_instance_create2` instead.
 */
```

I like this because it's a little more human readable and will hopefully be surfaced nicely by IDE tools (many of which understand Doxygen syntax).

The version number here is perhaps less machine-friendly than @Bromeon's suggestion, but if we adhere to this format strictly (and we'd need to be sure to adhere to any format strictly, including the previous suggestion) it can still be extracted by a parser.

Sauermann replied:
That sounds good. I have made the necessary documentation-comment changes.

akien-mga commented on tests/core/object/test_object.h:
```diff
@@ -36,6 +36,8 @@
 #include "core/object/object.h"
 #include "core/object/script_language.h"
 
+#include "modules/gdscript/gdscript.h"
```
This breaks encapsulation. Code outside `modules/gdscript/` shouldn't depend on that module.

So the relevant GDScript test should go in `modules/gdscript/tests/`.

Sauermann replied:
Moving the tests to `modules/gdscript/tests` is more tricky than initially expected.
For now I have removed the tests, that verify the correct notification order between native classes and GDScript, so that encapsulation is not broken.

akien-mga replied:
Alternatively, since this is test code we don't have to be too strict about itzl, so you could make the whole test and the include depend on `MODULE_GDSCRIPT_ENABLED`.

Sauermann replied:
Thanks, I will keep that in mind, if I don't manage to place the tests in `modules/gdscript`.

akien-mga reviewed (approved):
Looks good to me.

akien-mga commented:
Thanks!

maiself commented on core/extension/gdextension_interface.cpp:
```diff
@@ -1041,7 +1041,43 @@ static void gdextension_ref_set_object(GDExtensionRefPtr p_ref, GDExtensionObjec
+
+	ScriptInstanceExtension *script_instance_extension = memnew(ScriptInstanceExtension);
+	script_instance_extension->instance = p_instance_data;
+	script_instance_extension->native_info = &info_2;
```
This is taking the address of a temporary, which is causing crashes when the functions are called later.

dsnopek replied:
Ack, right because unlike `classdb_register_extension_class` we're not copying the data into a new internal struct, we're directly referring to the struct provided by the GDExtension. We should probably not do that - it'd probably be better to copy that data into an instance of the struct on the Godot side...

maiself replied:
Copying the data might be better in general. This is per instance tho, so either we duplicate the memory for each instance, or somehow create a mapping...

Another solution for this specific case is to add another class with a virtual override Godot side:

- `ScriptInstance`
- `ScriptInstanceExtensionDeprecated` (add this)
- `ScriptInstanceExtension`

Then have `gdextension_script_instance_create[2]` create the correct class.

dsnopek replied:
Here's a proposed fix that copies the data: https://github.com/godotengine/godot/pull/81205

> This is per instance tho, so either we duplicate the memory for each instance, or somehow create a mapping...

Hm, yeah, I suppose this does make each instance take up 23 pointers more memory... This is a trickier problem than I thought at first!

Sauermann replied:
Sorry about that. Looks like I don't have a good intuition for these kind of problems.

dsnopek replied:
> Another solution for this specific case is to add another class with a virtual override Godot side

Yeah, I worry about the scalability of this approach: what happens as we make more changes to this struct? (We already have a change queued up for https://github.com/godotengine/godot/issues/81119)

I'd like to avoid the compatibility code continuing to require maintenance and grow in complexity over time. Copying is at least simple and straight forward. But copying as done in my current PR is always a performance and memory hit, even on the new code path. What if the hit was only on the compatibility path, maybe that's an OK trade-off?

Directly relates this comment on the PR: https://github.com/godotengine/godot/pull/81205#issuecomment-1701686152

dsnopek replied:
> Sorry about that. Looks like I don't have a good intuition for these kind of problems.

No worries - several other people (including me) read through your code without noticing it :-)

Gallilus commented:
Hi running in to 
![image](https://github.com/godotengine/godot/assets/20573784/c6a8ec43-56ac-494d-955f-5cec5e0b4c46)
So want to ask some questions.

Currently, I am using a  GDExtensionScriptInstanceInfo2 without binding any methods

Gallilus commented on core/object/script_language_extension.h:
```diff
@@ -630,7 +630,11 @@ VARIANT_ENUM_CAST(ScriptLanguageExtension::CodeCompletionLocation)
-	const GDExtensionScriptInstanceInfo *native_info;
+	const GDExtensionScriptInstanceInfo2 *native_info;
+	struct {
+		GDExtensionClassNotification notification_func;
```
Why not GDExtensionScriptInstanceNotification?

dsnopek replied:
Ack, yes, you're right, this should be `GDExtensionScriptInstanceNotification`. Luckly, the C types of the arguments are equivalent, so I don't think this breaks anything, but it should be fixed.

Sauermann replied:
Is this just a straightforward rename from `GDExtensionScriptInstanceInfo2` to `GDExtensionScriptInstanceNotification`? In that case I can create a PR. Not sure about other implications.

dsnopek replied:
Yep, it should just be changing the type

Gallilus commented on core/extension/gdextension_interface.cpp:
```diff
@@ -1041,7 +1041,43 @@ static void gdextension_ref_set_object(GDExtensionRefPtr p_ref, GDExtensionObjec
+	ScriptInstanceExtension *script_instance_extension = memnew(ScriptInstanceExtension);
+	script_instance_extension->instance = p_instance_data;
+	script_instance_extension->native_info = &info_2;
+	script_instance_extension->deprecated_native_info.notification_func = p_info->notification_func;
```
Is it possible that `if not (p_info->notification_func) {script_instance_extension->deprecated_native_info.notification_func = nullptr}` is needed?

dsnopek replied:
No, I think this is fine. If `p_info->notification_func` is `nullptr`, then `script_instance_extension->deprecated_native_info.notification_func` will get set to `nullptr` as well.

Gallilus commented:
It tries to access a notification function when a .lr script gets attached. Only I have not defined a function jet so it should skip it. Not?
![image](https://github.com/godotengine/godot/assets/20573784/996197c8-5490-4109-9b82-832d9fcaf372)



code [here](https://github.com/Gallilus/godot-linker-language/blob/cd1d1f47abcd090356d7fa39e797920635a73e50/src/linker-script.cpp#L29-L59)
```
void *godot::LinkerScript::_instance_create(Object *for_object) const {
    Ref<LinkerScript> top = Ref<LinkerScript>(this);
    while (top->base.is_valid()) {
        top = top->base;
        ERR_PRINT("LinkerScript::_instance_create() " + String(top->_save_path));
    }


    return create_instance(nullptr, 0, for_object, Object::cast_to<RefCounted>(for_object) != nullptr);
}


GDExtensionScriptInstancePtr godot::LinkerScript::create_instance(const Variant **p_args, int p_argcount, Object *p_owner, bool p_is_ref_counted) const {
    
    /* STEP 1, CREATE */
    LinkerInstance *instance = memnew(LinkerInstance);
    instance->_base_ref_counted = p_is_ref_counted;
    instance->_script = Ref<Script>(this);
    instance->_owner = p_owner;
    instance->_owner_id = p_owner->get_instance_id();

    /* STEP 2, INITIALIZE AND CONSTRUCT */
    instance->initialize();

    return internal::gdextension_interface_script_instance_create2 (
        &LinkerInstanceGlue::INSTANCE_INFO,
        instance
    );
}
```

dsnopek commented:
@Gallilus:

How is `LinkerInstanceGlue::INSTANCE_INFO` initialized? All I could find in the code was the declaration [here](https://github.com/Gallilus/godot-linker-language/blob/cd1d1f47abcd090356d7fa39e797920635a73e50/src/linker-instance.cpp#L10) (where it's not initialized to any value) and [this](https://github.com/Gallilus/godot-linker-language/blob/cd1d1f47abcd090356d7fa39e797920635a73e50/src/linker-instance.h#L44) (which has all the assignments commented out).

In C++, memory isn't initialized to zero by default. So, if you aren't setting anything, `LinkerInstanceGlue::INSTANCE_INFO.notification_func` isn't `nullptr`, it's just some random value that happened to be in memory. :-)

Gallilus commented:
Found the problem, I think.
I now initialised all `INSTANCE_INFO` functions to `nullptr`.
This did not help to prevent crashes on `deprecated_native_info.notification_func` so set that to `nullptr` and it seems to work.
